[Ju Can] iP

Ju-Can opened the pull request.

Ju-Can committed:
Lvl 0 Greet

Ju-Can committed:
Completed level 2

Ju-Can committed:
level-2 create list

Ju-Can committed:
Level-1 completed

Ju-Can committed:
Level-2 completed

Ju-Can committed:
Completed Level-3

Ju-Can committed:
update code to follow coding standard

Ju-Can committed:
update code to follow coding standard

Ju-Can committed:
Partial Level-4 with buggy Event class

Event is abstract, cannot be instantiated

Ju-Can committed:
Complete code for Level 4

Ju-Can committed:
Improved code quality and minor implementation improvement

error message added when markTask() and unmarkTask() user input index goes out of bound

Richardtok commented on src/main/java/Duke.java:
```diff
@@ -19,59 +23,96 @@ public static void main(String[] args) {
+        }
+    }
+
+    public static void createTask(String input) {
```
You could have made a class for Task. Why did you do it this way instead of making a class?

Richardtok commented on src/main/java/Duke.java:
```diff
@@ -19,59 +23,96 @@ public static void main(String[] args) {
+        System.out.println(line + "added: " + t.getDescription() + '\n' + line);
+    }
+
+    public static void createTodo(String input) {
```
Why did you choose to do it this way, why not make a class for todo?

Richardtok commented on src/main/java/Duke.java:
```diff
@@ -19,59 +23,96 @@ public static void main(String[] args) {
+        System.out.println("Now you have " + countTask + " task(s) in the list.\n" + line);
+    }
+
+    public static void createEvent(String input) {
```
WHy did you do it this way why not create a class for event?

Richardtok commented on src/main/java/Duke.java:
```diff
@@ -19,59 +23,96 @@ public static void main(String[] args) {
-                "Type 'mark' and task event index to mark the task as done and type 'unmark'and index to undo the task.\n" +
+                "Type 'mark' and task event index to mark the task as done " +
+                "and type 'unmark' and index to undo the task.\n" +
                 line;
```
I like what you have done as the description of what the bot can do is very clear!

linuspuah commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,118 @@
+    }
+
+    public static void markTask(String input) {
+        int taskIndex = Integer.parseInt(input.substring(4).trim());
```
What is the 4 here representing, if possible can create a variable name for the number instead of leaving it as a magic number.

linuspuah commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,118 @@
 public class Duke {
+    public static String line = "____________________________________________________________\n";
+    public static int countTask = 0;
+    public static Task[] tasks = new Task[100];
```
Can create a constant MaxTasks to represent the 100 instead of leaving it as a magic number

linuspuah commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+public class Task {
+    protected String description;
```
Maybe a clearer name for description would be taskDescription. (applicable change to other classes as well)

linuspuah commented on src/main/java/ToDo.java:
```diff
@@ -0,0 +1,10 @@
+    }
+
+    @Override
+    public String printTask(){
+        return "[T]" + super.printTask();
+    }
```
Maybe the naming of this method could be more appropriately changed to getTaskIdentity since the actual method itself is not printing anything.

linuspuah commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,118 @@
+    }
+
+    public static void createTask(String input) {
+        Task t = new Task(input);
```
Can try to rename the new task t to newTask as t fails to provide ample description of the variable

Ju-Can committed:
implemented testing and some moderation of output format

Ju-Can committed:
update format

Ju-Can committed:
first implementation of error detection

Ju-Can committed:
catch more errors

catch error for both empty input for deadline/event and no specific time

Ju-Can committed:
Error Handling

no date/time for deadline and event + wrong order for start and ending time for event handled

Ju-Can committed:
Merge branch 'branch-Level-5'

Ju-Can committed:
Add package duke.task

Ju-Can committed:
Merge branch 'branch-A-Packages'

Ju-Can committed:
added delete task feature

Ju-Can committed:
write to file and half of reading in

Ju-Can committed:
save taskList to and read from local file

Ju-Can committed:
Merge branch 'branch-Level-6'

Ju-Can committed:
merge level-7

Ju-Can committed:
adjust format for level-7

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,261 @@
+
 public class Duke {
+    public static String line = "____________________________________________________________\n";
+    //    public static int countTask = 0;
+//    public static Task[] tasks = new Task[100];
```
Consider removing these lines of code if you no longer plan on/require them, in order to improve code readability.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,261 @@
+        for (Task t : tasks) {
+            String completeTaskDescription = t.printTask();
+            String taskType = completeTaskDescription.substring(1, 2);
+            switch (taskType) {
+            case "T":
+                fw.write(taskType + " | " + t.getStatusIcon() + " | " + t.getDescription() +
+                        System.lineSeparator());
+                break;
+            case "D":
+                fw.write(taskType + " | " + t.getStatusIcon() + " | " + t.getDescription() + " | " +
+                        ((Deadline) t).getDeadline() + System.lineSeparator());
+                break;
+            case "E":
+                fw.write(taskType + " | " + t.getStatusIcon() + " | " + t.getDescription() + " | " +
+                        ((Event) t).getStartTime() + " | " + ((Event) t).getEndTime() + System.lineSeparator());
+                break;
+            default:
+            }
```
Perhaps you can consider abstracting this block of code to a separate function in order to minimise deeply-nested code and improve code readability.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,261 @@
+                break;
+            }
+            try {
+                if (userInput.equals("bye")) {
+                    System.out.println("Thank you for using Duke. Hope to see you soon!");
+                    break;
+                } else if (userInput.equals("list")) {
+                    printTaskList();
+                } else if (userInput.startsWith("mark")) {
+                    markTask(userInput);
+                } else if (userInput.startsWith("unmark")) {
+                    unmarkTask(userInput);
+                } else if (userInput.startsWith("todo")) {
+                    try {
+                        createTodo(userInput);
+                    } catch (IndexOutOfBoundsException e) {
+                        System.out.println("Description of a todo cannot be empty. Adding todo failed.\n" + line);
+                    }
+                } else if (userInput.startsWith("deadline")) {
+                    try {
+                        createDeadline(userInput);
+                    } catch (StringIndexOutOfBoundsException e) {
+                        System.out.println("Insufficient input for deadline. Adding deadline failed.\n" + line);
+                    }
+                } else if (userInput.startsWith("event")) {
+                    try {
+                        createEvent(userInput);
+                    } catch (IndexOutOfBoundsException e) {
+                        System.out.println("Insufficient input for event. Adding event failed.\n" + line);
+                    }
+                } else if (userInput.startsWith("delete")) {
+                    deleteTask(userInput);
+                } else {
+                    throw new TaskTypeException();
+                }
```
Likewise, you might want to consider abstracting this section of your progarm as it is quite long and contains deeply-nested code.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,261 @@
+    }
+
+    public static void createEvent(String input) throws IndexOutOfBoundsException, EventTimingException {
+        if (!input.contains("/from") || !input.contains("/to") || input.indexOf("/from") + 6 > input.length()
+                || input.indexOf("/to") + 4 > input.length()) {
+            throw new StringIndexOutOfBoundsException();
+        } else if (input.indexOf("/from") > input.indexOf("/to")) {
+            throw new EventTimingException();
```
I noticed you make several references to your command specifiers such as `/from` and `/to`. You can perhaps consider assigning these values to a variable then referencing them in order to minimise the use of magic literals.

exetr reviewed (commented):
Great progress on your iP. I have left some comments regarding potential areas for improvment. Specifically, I suggest taking a look through `Duke.java` and minimising the use of magic literals/strings.

Ju-Can committed:
Refactor code

Ju-Can committed:
Find tasks using keyword

Ju-Can committed:
Merge pull request #1 from Ju-Can/branch-Level-9

Merge level-9

Ju-Can committed:
Added JavaDoc

Ju-Can committed:
Updated welcome message to include a dummy help page

Ju-Can committed:
update docs/README

Ju-Can committed:
Add help page

Ju-Can committed:
remove help page

Ju-Can committed:
update typo in docs/readme

Ju-Can committed:
Update welcome message
